src/spikeinterface/core/base.py

```diff
@@ -846,43 +846,63 @@ def save_to_memory(self, sharedmem=True, **save_kwargs) -> "BaseExtractor":
+    def save_to_folder(
+        self,
+        name: str | None = None,
+        folder: str | Path | None = None,
+        overwrite: str = False,
+        verbose: bool = True,
+        **save_kwargs,
+    ):
         """
-        Save extractor to folder.
+        Save the extractor and its data to a folder.
 
-        The save consist of:
-          * extracting traces by calling get_trace() method in chunks
-          * saving data into file (memmap with BinaryRecordingExtractor)
-          * dumping to json/pickle the original extractor for provenance
-          * dumping to json/pickle the cached extractor (memmap with BinaryRecordingExtractor)
+        This method extracts trace data, saves it to a file (using the memory-mapped
+        BinaryRecordingExtractor format), and stores both the original extractor's provenance
+        and the cached extractor's metadata in JSON format.
 
-        This replaces the use of the old CacheRecordingExtractor and CacheSortingExtractor.
+        The folder's final location and name can be specified in a couple of ways ways:
 
-        There are 2 option for the "folder" argument:
-          * explicit folder: `extractor.save(folder="/path-for-saving/")`
-          * explicit sub-folder, implicit base-folder : `extractor.save(name="extarctor_name")`
-          * generated: `extractor.save()`
+        1. Explicitly providing the full path:
+        ```
+        extractor.save_to_folder(folder="/path/to/save/")
+        ```
 
-        The second option saves to subfolder "extractor_name" in
-        "get_global_tmp_folder()". You can set the global tmp folder with:
-        "set_global_tmp_folder("path-to-global-folder")"
+        2. Providing a subfolder name, with the base folder being determined automatically:
+        ```
+        extractor.save_to_folder(name="my_extractor_data")
+        ```
+        In this case, the data is saved in a subfolder named "my_extractor_data"
+        within the global temporary folder (set using `set_global_tmp_folder`). If no
+        global temporary folder is set, one will be generated automatically.
 
-        The folder must not exist. If it exists, remove it before.
+        3. If neither `name` nor `folder` is provided, a random name will be generated
+        for the subfolder within the global temporary folder.
 
         Parameters
         ----------
-        name: None str or Path
-            Name of the subfolder in get_global_tmp_folder()
-            If "name" is given, "folder" must be None.
-        folder: None str or Path
-            Name of the folder.
-            If "folder" is given, "name" must be None.
-        overwrite: bool, default: False
-            If True, the folder is removed if it already exists
+        name : str , optional
+            The name of the subfolder within the global temporary folder. If `folder`
+            is provided, this argument must be None.
+        folder : str or Path, optional
```

Comment:
folder : Optional[str | Path], default: None

or

folder: str | Path | None , default: None

would be my two preferences. Using `,optional` isn't the `Optional` I'm okay with.

sorry this is tricky. I think my own confusion came from the fact the way I was trying to describe Optional how I thought it would work:

I never expected a function to do `(argument: Optional[str]= 'home_folder')` instead of `(argument: Optional[str] = None)`. But technically you use Optional for both. It just seems like poor design to accept None if it isn't the default argument for me.

Basically numpy optional is to indicated required vs optional arguments, which to me already seems clear based on the function parameters.

```
func(x, y=1)
"""x: float
     y, optional
"""
```

Reply:
```
I never expected a function to do (argument: Optional[str]= 'home_folder') instead of (argument: Optional[str] = None). But technically you use Optional for both. It just seems like poor design to accept None if it isn't the default argument for me.
```

I agree with this, I would not like to use optional in that case.

I am more confused about numpy convention, can you link it?
My own preference would be to use default=value if there is a default that is not None  and optional to mean that you don't have to specify it.

Reply:
https://numpydoc.readthedocs.io/en/latest/format.html

Here's their style guide. They don't try to force you to use optional, but if you use vscode numpydoc style then it autofills in the optional. So it is a recommendation that I think the auto-docstring complete takes too far.

Reply:
So I shouldn't blame numpydoc, but instead I should blame our overreliance on tooling :)

Reply:
OK, so numpy doc optional is more expansive than the one in typing.

So, what do you think of using the following option:
* optional only, when is optional in the typing sense (argument=None)
* default, when there is a default value that is not None.

It seems consistent with their docs as they say
> If the default value would not be used as a value, optional is preferred. These are all equivalent:

So, the docstring here is like this:
values that are optional in the typing sense are marked as optional
values that have a default are marksed as default.

What do you think?

Reply:
Can I mark this as resolved, @zm711? I moved the discussion to https://github.com/SpikeInterface/SpikeInterface-Hackathon-Edinburgh-May24/issues/29

Reply:
@zm711 ?

Reply:
Thanks, man.